### PR TITLE
feat: ask for environment branching

### PR DESCRIPTION
**Pull request by Rusty-Beard**

### CODE REVIEW

**Comment by Rusty-Beard** [comment]
depends on https://github.com/8base/cli/pull/131

File: e2e-scaffold/setup/testEnvironment.js [hunk 1]
```diff
@@ -1,17 +1,17 @@
...
-const NodeEnvironment = require("jest-environment-node");
-const fs = require("fs");
+const jestNodeEnvironment = require("jest-environment-node");
+const fs = require("fs-extra");
 const os = require("os");
 const path = require("path");
 
 const DIR = path.join(os.tmpdir(), "jest_puppeteer_global_setup");
-
+const NodeEnvironment = jestNodeEnvironment.default || jestNodeEnvironment;
```
Comment (Rusty-Beard) [inline e2e-scaffold/setup/testEnvironment.js:8]: workaround to correctly initialize class

File: src/engine/commands/deploy/index.ts [hunk 2]
```diff
@@ -1,20 +1,15 @@
...
 export default {
   command: 'deploy',
-  handler: async (params: any, context: Context) => {
+  handler: async (params: { plugins?: string[]; functions?: string[]; mode: DeployModeType }, context: Context) => {
     context.initializeProject();
 
-    if (params['validate_schema']) {
-      GraphqlController.validateSchema(context.project);
-    }
-
```
Comment (Rusty-Beard) [inline src/engine/commands/deploy/index.ts:17]: don't have such arg any more

File: src/engine/commands/init/index.ts [hunk 3]
```diff
@@ -99,43 +95,17 @@ export default {
...
-        choices: [
-          {
-            title: '<New Workspace>',
-            value: 'NEW_WORKSPACE',
-          },
-          ...workspaces.map((workspace: any) => ({
-            title: workspace.name,
-            value: workspace.id,
-          })),
-        ],
```
Comment (Rusty-Beard) [inline src/engine/commands/init/index.ts:111]: not actual any more

File: src/engine/commands/login/webLogin.ts [hunk 4]
```diff
@@ -1,16 +1,15 @@
 import { Context } from '../../../common/context';
-import * as cuid from 'cuid';
+import * as cuid from '@paralleldrive/cuid2';
```
Comment (s1owjke) [inline src/engine/commands/login/webLogin.ts:2]: It's probably better to use the same package as on the server

File: src/engine/commands/logs/index.ts [hunk 5]
```diff
@@ -181,28 +177,28 @@ const readLogs = async (functionName: string, context: Context) => {
...
   }
 };
 
 export default {
-  command: 'logs [name]',
-  handler: async (params: any, context: Context) => {
-    if (params.n > 100) {
-      params.n = 100;
+  command: 'logs <name>',
+  handler: async (params: { name: string; num: number; tail?: boolean }, context: Context) => {
```
Comment (s1owjke) [inline src/engine/commands/logs/index.ts:188]: Let's create `LogParams` type here and let's do the same for other commands

File: src/engine/commands/init/index.ts [hunk 6]
```diff
@@ -14,36 +14,40 @@ import { Colors } from '../../../consts/Colors';
...
 
-  handler: async (params: any, context: Context) => {
-    const { functions, empty, syntax, mocks, silent } = params;
+  handler: async (params: InitParams, context: Context) => {
+    const { name, functions, empty, syntax, mocks, silent } = params;
 
     let { workspaceId, host } = params;
 
-    const [, projectName] = _.castArray(params._);
+    const projectName = name || path.basename(context.config.rootExecutionDir);
```
Comment (s1owjke) [inline src/engine/commands/init/index.ts:49]: Previous version was more readable

```
const project = name
  ? {
      fullPath: path.join(context.config.rootExecutionDir, name),
      name,
    }
  : {
      fullPath: context.config.rootExecutionDir,
      name: path.basename(context.config.rootExecutionDir),
    };
```

File: src/engine/commands/import/index.ts [hunk 7]
```diff
@@ -1,18 +1,21 @@
...
+    params: { file: string; schema: boolean; data: boolean; workspace: string; d?: boolean },
+    context: Context,
+  ) => {
     let schema;
 
-    if (fs.existsSync(params.file)) {
+    if (await fs.exists(params.file)) {
       try {
-        schema = JSON.parse(fs.readFileSync(params.file, 'utf8'));
+        schema = JSON.parse(await fs.readFile(params.file, 'utf8'));
```
Comment (s1owjke) [inline src/engine/commands/import/index.ts:18]: In `fs-extra` you can use `readJson` instead

File: src/engine/commands/environment/commands/branch.ts [hunk 8]
```diff
@@ -28,12 +60,29 @@ export default {
...
         choices: Object.values(MigrateMode),
+        requiresArg: true,
+      })
+      .option('force', {
+        alias: 'f',
+        describe: translations.i18n.t('environment_branch_force_describe'),
       }),
 };
+
+const isEnvironmentLimitReached = (e: any): string => {
```
Comment (s1owjke) [inline src/engine/commands/environment/commands/branch.ts:79]: Let's create utility to check response error codes

File: src/engine/commands/environment/commands/branch.ts [hunk 9]
```diff
@@ -28,12 +60,29 @@ export default {
...
+  const err =
+    !_.isEmpty(e.response?.errors) &&
+    e.response.errors.find((err: any) => err.code === errorCodes.BillingPlanLimitWarningCode);
+  if (!err || !err.message) {
+    throw e;
```
Comment (s1owjke) [inline src/engine/commands/environment/commands/branch.ts:84]: What will the user see in the console?

File: e2e/tests/commands/invoke-local/invokeWithParameters.test.ts [hunk 10]
```diff
@@ -5,12 +5,12 @@ import {
   prepareTestEnvironment,
   RunCommand,
 } from '../../../utils';
+import { jest } from '@jest/globals';
```
Comment (s1owjke) [inline e2e/tests/commands/invoke-local/invokeWithParameters.test.ts:8]: Do we need this?

File: src/common/utils.ts [hunk 11]
```diff
@@ -142,4 +144,15 @@ export namespace Utils {
         };
       }
     };
+
+  export const checkHttpResponse = async (
```
Comment (s1owjke) [inline src/common/utils.ts:148]: Like in go, but let's use throw here instead? It should make code easier

File: src/engine/controllers/projectController.ts [hunk 12]
```diff
@@ -209,16 +207,17 @@ export class ProjectController {
...
       throw new InvalidConfiguration(pathToYmlConfig, ex.message);
     }
   }
 
-  private static saveConfigFile(context: Context, config: Object, projectPath?: string, silent?: boolean): any {
+  private static saveConfigFile(context: Context, config: Object, projectPath?: string, silent?: boolean): void {
     const pathToYmlConfig = projectPath ? path.join(projectPath, '8base.yml') : StaticConfig.serviceConfigFileName;
 
-    fs.writeFileSync(pathToYmlConfig, yaml.safeDump(config));
+    const dump = yaml.dump(config);
```
[the remark this answers is not in the document]
Comment (s1owjke) [reply src/engine/controllers/projectController.ts:216]: Was changed in `js-yaml@4`

File: src/engine/controllers/projectController.ts [hunk 13]
```diff
@@ -209,16 +207,17 @@ export class ProjectController {
     }
 
     try {
-      return yaml.safeLoad(fs.readFileSync(pathToYmlConfig, 'utf8'));
+      return yaml.load(fs.readFileSync(pathToYmlConfig, 'utf8'));
```
[the remark this answers is not in the document]
Comment (s1owjke) [reply src/engine/controllers/projectController.ts:207]: Was changed in `js-yaml@4`

**Pull request merged**